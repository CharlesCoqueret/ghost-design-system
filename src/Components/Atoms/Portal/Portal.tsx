import { PropsWithChildren, ReactElement, useEffect, useRef } from 'react';
import { createPortal } from 'react-dom';

export interface IPortalProps {
  rootId?: string;
}

const Portal = (props: PropsWithChildren<IPortalProps>): ReactElement => {
  const { children, rootId } = props;
  const target = useRef<HTMLDivElement>();

  useEffect(() => {
    let container = document.getElementById(rootId || Portal.defaultProps.rootId);
    if (!container) {
      container = document.createElement('div');
      container.setAttribute('id', rootId || Portal.defaultProps.rootId);
      document.body.appendChild(container);
    }

    if (target.current) {
      container.appendChild(target.current);
    }

    return () => {
      if (target.current) {
        target.current.remove();
      }
      if (container && container.childNodes.length === 0) {
        container.remove();
      }
    };
  }, [rootId]);

  if (!target.current) {
    target.current = document.createElement('div');
  }

  return createPortal(children, target.current);
};

Portal.defaultProps = {
  rootId: 'root-portal-id',
};

export default Portal;
