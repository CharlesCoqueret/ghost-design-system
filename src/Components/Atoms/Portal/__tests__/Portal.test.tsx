import React from 'react';
import { render } from '@testing-library/react';

import Portal from '../Portal';

afterEach(() => {
  jest.useRealTimers();
});

describe('Portal Component', () => {
  it('Portal renders', () => {
    const container = render(
      <Portal rootId='root-id'>
        <div />
      </Portal>,
    );

    expect(container.container).toMatchSnapshot();

    container.rerender(<Portal rootId='root-id' />);

    expect(container.container).toMatchSnapshot();

    container.unmount();

    expect(container.container).toMatchSnapshot();
  });

  it('Portal renders without rootId', () => {
    const container = render(
      <Portal>
        <div />
      </Portal>,
    );

    expect(container.container).toMatchSnapshot();

    container.rerender(<Portal />);

    expect(container.container).toMatchSnapshot();

    container.unmount();

    expect(container.container).toMatchSnapshot();
  });
});
