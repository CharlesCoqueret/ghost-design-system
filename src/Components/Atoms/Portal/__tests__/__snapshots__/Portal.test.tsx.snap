// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal Component Portal renders 1`] = `
<div>
  <div>
    <div />
  </div>
</div>
`;

exports[`Portal Component Portal renders 2`] = `
<div>
  <div />
</div>
`;

exports[`Portal Component Portal renders 3`] = `<div />`;

exports[`Portal Component Portal renders without rootId 1`] = `
<div>
  <div>
    <div />
  </div>
</div>
`;

exports[`Portal Component Portal renders without rootId 2`] = `
<div>
  <div />
</div>
`;

exports[`Portal Component Portal renders without rootId 3`] = `<div />`;
